### ISSUE TITLE
Paddle API v4 proposal

### ISSUE THREAD

**Issue by helinwang** [PaddlePaddle/Paddle#10152]
Pytorch uses Python to call each operator, the following API uses Python to call each compiled function.

### fit_a_line.py
```Python

class FitALine(fluid.Program):
  DEFAULT_READER_FILE_PATH =  './data.recordio'
  DEFAULT_READER_BATCH_SIZE = 128 

  # @network decorator will be used by the compiler to generate a
  # ProgramDesc block.  It can optionally take inputs, which represents a
  # mapping from the input vars to vars in the block
  @network()
  def train_step(self):
    reader = fluid.batch_reader(file=DEFAULT_READER_FILE_PATH,
                                               batch_size=DEFAULT_READER_BATCH_SIZE,
                                               shape=[[13], [1]], 
                                               dtype=['float32', 'float32'],
                                               format='recordio')
    x, y = self.reader.next_item()
    with fluid.var_scope('prediction'):
      # Since we want to be able to access the same weights and bias during
      # inference, we need to namespace the variables.
      # fluid.var_scope block guard will create a new UniqueNameGenerator when
      # we enter the block, and rollback to the previous UniqueNameGenerator when
      # the block exits.
      y_predict = fluid.layers.fc(input=x, size=1, act=None)

    cost = fluid.layers.square_error_cost(input=y_predict, label=y)
    avg_cost = fluid.layers.mean(cost, name='avg_cost')
    sgd_optimizer = fluid.optimizer.SGD(learning_rate=0.001)
    sgd_optimizer.minimize(avg_cost)
    return avg_cost

  @network("x")
  def infer(self):
    x = fluid.layers.data(name='x', shape=[13], dtype='float32')
    with fluid.var_scope('prediction'):
      return fluid.layers.fc(input=x, size=1, act=None)
```

### main.py
```Python
fit_a_line = FitALine(batch_size=256).Compile()

while i in range(1000):
  avg_cost = fit_a_line.train_step()

y_results = fit_a_line.infer([3,4,6,3,5,7,8,6,5,4,1,5,8])
```

#### Trainer
```
> TRAINING_ROLE=TRAINER python main.py --distributed main=train_step
```
#### PServer
```
> TRAINING_ROLE=PSERVER paddle run train.py --distributed main=train_step
```

**Comment by cs2be** [comment]
this looks good.  one question, when is the reader initialized?  when the first iteration is ran?  we probably need to implicitly do this, maybe by creating a static or shared reader?

**Comment by varunarora** [comment]
How about:

### main.py
```Python
from fit_a_line import FitALine

# fluid.program = fluid.compile() and fluid.init()
trainer = FitALine().train_step(batch_size=256)

while result in trainer.run(1000):
  cost = result['avg_cost']

y_results = FitALine().infer([3,4,6,3,5,7,8,6,5,4,1,5,8])
```

**Comment by cs2be** [comment]
Hi @wangkuiyi,

1) This is a good point, we are still debating on this, since transpiler may need to know what method the main program is in.  Currently for transpilation step, we let users define what the main program method is.

2) We are thinking that each method will be compiled to a ProgramDesc.  This will allow the users to run them in any order they want.  We still need to refine this idea.

3) This is a good point, I'll discuss with team to see what a good solution for this is.

**Comment by panyx0718** [comment]
I think this proposal looks mostly good!

One question:
```python
> TRAINING_ROLE=TRAINER python main.py --distributed main=train_step
```
How does the Compile() logic magically knows how to transpile a program into
trainer part and ps part?

In more detail, how does the transpiler know which operators and variables should
be placed in trainer and others be placed in parameter servers?

**Comment by helinwang** [comment]
@panyx0718 thanks for the review!

> How does the Compile() logic magically knows how to transpile a program into
trainer part and ps part?
 In more detail, how does the transpiler know which operators and variables should
be placed in trainer and others be placed in parameter servers?

Our current transpiler implementation does "magically" make everything work. Indeed, it is hacky because a lot of assumptions has been made. To improve, the transpiler needs to know more information, such as "which operators belongs to the optimizers", "which operators belong to the gradient calculating pass", "which operators are explicitly defined by the user". I think we should store these information in the program desc.

Do you think the API makes sense, regardless of the current transpiler implementation (or assuming we have implemented it in the non-hacky way)?

**Comment by abhinavarora** [comment]
Can we change `fluid.var_scope ` to `fluid.scope` because `Variable`  is a C++ concept in Paddle and should not be exposed to Python users

**Comment by panyx0718** [comment]
@helinwang 

I think the skeleton looks fine.
Though I still don't know how to achieve the following with "--distribute" option
> the transpiler needs to know more information, such as "which operators belongs to the optimizers", "which operators belong to the gradient calculating pass", "which operators are explicitly defined by the user"

**Comment by panyx0718** [comment]
@PaddleCI 

Other Paddle team members can take a look at this proposal.

**Comment by jacquesqiao** [comment]
Can @network() support training multiple networks, like GAN?

**Comment by helinwang** [comment]
@jacquesqiao thanks for reviewing! Yes I think so. Multiple class methods annotated by `@network()` share variable by using block guard `with fluid.var_scope("shared_name_scope")`.
The program generated by `fluid.Program.Compile` owns the scope, which contains the parameters of however many networks.

**Comment by shanyi15** [comment]
您好，此issue在近一个月内暂无更新，我们将于今天内关闭。若在关闭后您仍需跟进提问，可重新开启此问题，我们将在24小时内回复您。因关闭带来的不便我们深表歉意，请您谅解~感谢您对PaddlePaddle的支持! 
 Hello, this issue has not been updated in the past month. We will close it today for the sake of other user‘s experience. If you still need to follow up on this question after closing, please feel free to reopen it. In that case, we will get back to you within 24 hours. We apologize for the inconvenience caused by the closure and thank you so much for your support of PaddlePaddle Group!
